e2eTests/.fructose/jest-setup.js

```diff
@@ -3,46 +3,97 @@ import fructose from "@times-components/fructose/setup";
 
+const isPortTaken = port =>
+  new Promise(resolve => {
+    portscanner.checkPortStatus(port, "127.0.0.1", (error, status) => {
+      // Status is 'open' if currently in use or 'closed' if available
+      const isTaken = status === "open";
+      resolve(isTaken);
+    });
+  });
+
+const checkIfPortTaken = x =>
+  new Promise(resolve => {
+    const event = new EventEmitter();
+    event.on("taken", taken => {
+      resolve(taken);
+    });
+    let taken;
+    let repeatTimes = x;
+    const interval = setInterval(async () => {
+      repeatTimes -= 1;
+      taken = await isPortTaken(3000);
+      if (taken) {
+        event.emit("taken", true);
+        clearInterval(interval);
+      }
+      if (!taken && repeatTimes === 0) {
+        event.emit("taken", false);
+        clearInterval(interval);
+      }
+    }, 1000);
+  });
+
 let appium;
 fructose.withComponent();
 
 beforeAll(async () => {
-  await fructose.hooks.setup();
-  if (process.env.ANDROID) {
-    appium = await new Promise(resolve => {
-      const proc = spawn("appium");
-      proc.stdout.on("data", d => {
-        if (d.toString("utf8").includes("started on 0.0.0.0:4723")) {
-          console.log(d.toString("utf8"));
-          resolve(proc);
-        }
+  if (process.env.ANDROID || process.env.IOS) {
+    await fructose.hooks.mobile.setup();
+    if (process.env.ANDROID) {
+      appium = await new Promise(resolve => {
+        const proc = spawn("appium");
+        proc.stdout.on("data", d => {
+          if (d.toString("utf8").includes("started on 0.0.0.0:4723")) {
+            resolve(proc);
+          }
+        });
       });
-    });
-    const options = {
-      desiredCapabilities: {
-        platformName: "Android",
-        platformVersion: "7.0",
-        deviceName: "Android Emulator",
-        app:
-          "/Users/rohanjanjua/workspace/News/fructose/e2eTests/android/app/build/outputs/apk/app-debug.apk"
-      },
-      host: "localhost",
-      port: 4723
-    };
-    global.driver = webdriverio.remote(options);
-    await global.driver.init();
-    await global.driver.waitForVisible('//*[@text="Fructose"]', 45000);
-  } else {
-    await detox.init(config.detox);
+      const options = {
+        desiredCapabilities: {
+          platformName: "Android",
+          platformVersion: "7.0",
+          deviceName: "Android Emulator",
+          app: path.join(
+            __dirname,
+            "../android/app/build/outputs/apk/app-debug.apk"
+          )
+        },
+        host: "localhost",
+        port: 4723
+      };
+      global.driver = webdriverio.remote(options);
+      await global.driver.init();
+      await global.driver.waitForVisible('//*[@text="Fructose"]', 45000);
+    } else {
+      await detox.init(config.detox);
+    }
+  } else if (process.env.WEB) {
+    await fructose.hooks.web.setup();
+    const taken = await checkIfPortTaken(10);
+    if (!taken) {
+      throw new Error("fructose app is not running");
```

Comment:
pretty sure this should stop test execution but double checking

Reply:
I don't think it actually does due to the method being async.
will change.